packages/sdk/src/across/rateModel.ts

```diff
@@ -0,0 +1,143 @@
+import { expectedRateModelKeys, RateModel } from "./constants";
+import { exists } from "../utils";
+import { ethers } from "ethers";
+
+// Each L1 token is mapped to an array of stringified rate models, ordered by the block height at which they were
+// published on-chain. This dictionary is used internally to fetch a rate model for a block height.
+type RateModelEventsDictionary = {
+  [l1TokenAddress: string]: { blockNumber: number; rateModel: string }[];
+};
+
+// Events should be reformatted into this shape to be used as input into methods in this file.
+export type RateModelEvent = {
+  blockNumber: number;
+  rateModel: string;
+  l1Token: string;
+};
+
+/**
+ * Helper method that returns parsed rate model from string, or throws.
+ * @param rateModelString Stringified rate model to parse.
+ * @returns Rate model object. Must conform to `expectedRateModelKeys` format.
+ */
+
+export const parseAndReturnRateModelFromString = (rateModelString: string): RateModel => {
+  const rateModelFromEvent = JSON.parse(rateModelString);
+
+  // Rate model must contain all keys in `expectedRateModelKeys`, and extra keys are OK.
+  for (const key in expectedRateModelKeys) {
+    if (!(key in Object.keys(rateModelFromEvent))) {
+      throw new Error(
+        `Rate model does not contain all expected keys. Expected keys: [${expectedRateModelKeys}], actual keys: [${Object.keys(
+          rateModelFromEvent
```

Comment:
I asked about this in another comment, but I think I explained it badly, so I'll give an example. Do you think we should throw if the on-chain rate model has an extra key `R3` that the bot's rate model computation doesn't need or use?

Right now, it think we just check that it has the keys we use, but we don't care if there are extras, but wouldn't an extra key indicate that our rate model computation is out of date or incompatible?

Reply:
> Do you think we should throw if the on-chain rate model has an extra key R3 that the bot's rate model computation doesn't need or use?

No i don't think an extra key matters, as long as the rate model can be computed using the keys it does have? I don't think this makes the rate model incompatible or out of date, this is not obvious to me

Reply:
Let's say we want to add an R3 to make this piecewise rate curve have an additional changepoint (2 instead of 1).

I'm assuming that means we would need to change the names of the _existing_ keys too to make sure that the old clients error out instead of performing the wrong computation right?

Or do you think we should include a version number key?

Basically, I think we should have _some_ mechanism to ensure that updates to rate models are recognized in old clients so they can throw errors instead of doing the wrong computation. How do you think we should do that?

Reply:
Ah i see your point. I'll just add the reverse check as well to guarantee all keys are exactly included and no extra ones